APIの全体像をOpenAPI Specに基づいて作成

a5ob7r opened the pull request:
# 概要

単純なオンラインストレージサービスという想定で必要になりそうなエンドポイントを書き出した。今後は、それぞれのエンドポイントの詳細を一つずつ詰めていくという流れで作業することを想定している。

# 参考資料

- OpenAPI Specification Version 3.1.0
  https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md

a5ob7r committed:
Add API overview with OpenAPI spec

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+  version: 0.0.0
+
+paths:
+  /signup:
```
REST の考えではエンドポイント (URL のパス部) はリソース名を表現します。
今回の設計ではリソース名ではなく振る舞い・操作を表しているので、この API は RESTful ではないですね。

RESTful ではない API であっても問題はないのだけど、

* Rails は REST の考えに従っていると書きやすい
* REST API はデファクトスタンダードだと思う

あたりから、エンドポイントを変えるかどうかちょっと迷いますね。

REST に従う場合、ユーザーの作成 API は `POST /users` となるのが妥当ですね。

a5ob7r replied:
この辺りは、参考にしたRailsチュートリアルのサンプルアプリケーションでこのようなルーティングを採用しており、それをそのまま踏襲したためにこのようにしています。この辺りは一般にどうするのがいいのでしょうか。あちらはWEBアプリケショーンでしたので、ある種適当というか、ある程度自由に記述しても問題なかった一方で、別の箇所でのコメントでもありましたが、WEB APIでは利用者に伝わるかどうかが重要とのことだったので、それでいうと、安易に別名を利用するのではなく、まずはRESTfulに忠実にそうべきなんですかね。

tk0miya replied:
開発者の裁量の範囲なので

* 恣意的な名前をつける
* REST の原則に従った名前をつける

というどちらを選択しても誤りではありません。
決めの問題だと思うので、好みで決めちゃってください。

ひとつだけ触れておくと、Web アプリケーションの場合は利用者が一般ユーザーであることに対して、API の場合はプログラムから呼ばれるので使う人間は開発者というのが違いでしょうか。

a5ob7r replied:
久しぶりに見直してみると、 `/user` エンドポイントがあるのにどうしてわざわざ `/signup` を用意するのかと、(自分のことながら)疑問に思いました。ですので、 `/signup` エンドポイントは `/user` エンドポイントに統合してしまおうと考えています。

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+paths:
+  /signup:
+    summary: Create a new user
+    get:
+      summary: Return a form to create a new user
```
金曜の夕会で会話しましたが、API ではフォームを扱うことはしません。
API クライアントがどういうふうに API を利用するかも分かっていない (UI を持っているかどうか、ユーザーの入力を受け付けるのかどうか) ですし、「フォームを返す」というのは現実的ではありません。

一般的には API 定義を提供することによって (どういうパラメータを受け付けるのかを宣言することによって) フォームの代わりとしていますね。

a5ob7r replied:
おっしゃる通りなので該当するエンドポイントは修正、というか削除します。

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+      summary: Return a form to create a new user
+    post:
+      summary: Create a new user using the posted information
+  /signin:
```
上のコメントと同様です。
サインインというのは操作であってリソースではないため、この名前はちょっと悩みどころです。
REST では `/session`, `/sessions`, `/tokens`, `/token` などで表現する気がしますね。
(セッションやトークンといった仮想リソースを使って、ログイン・ログアウト操作を表現します)

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+      summary: Return a form to log in as a user
+    post:
+      summary: Log in as a user
+    delete:
+      summary: Log out as a user
```
現状のまま `/signin` をエンドポイントにするのであれば、ログアウトは `/signout` というエンドポイントを用意したいですね。
`DELETE /signin` ってちょっと違和感ありませんか?

a5ob7r replied:
完全に同意です。参考にしたRailsチュートリアルのサンプルアプリでもそうしていました。なんで `/signin` に投げるようにしたのか、自分でも思い出せない。脳死だったのかな。 `/signout` に投げるよう修正します。

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+    patch:
+      summary: Update info of a logged-in user
+    delete:
+      summary: Log out as a user
```
質問: `DELETE /signin` との違いを知りたいです

a5ob7r replied:
あー、これはユーザを削除するためのAPIですね。 `summary` が間違っているので修正します。

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+    parameters:
+      - name: fileId
+  /profile:
+    get:
+      summary: Get info of a logged-in user
```
summary では "user" と読んでいますが、エンドポイントでは `/profile` なのは矛盾を感じました。意図して使い分けていますか?

a5ob7r replied:
今読んでみると、 `/user` の自然ですね。記述した当時は、 `GET` や `POST` でユーザの登録情報を更新するなら、対象としてはユーザそのものというよりプロフィールだな、みたいに考えたのかもしれません。あと、これは完全に僕の頭の中にしかない情報ですが、今回のAPIというかサービスでは、少なくとも現段階では、ユーザ同士は一切のやり取りが発生しないことになっています。なので、APIユーザは、ユーザではなく自分自身のみを認識すればいいので、あえてユーザというエンドポイントを用意する必要はないのでは、と感じたのかもしれません。ただ、説明ではユーザという用語をがっつり使っているので、結局意図していたわけではないというオチかもしれません。

a5ob7r replied:
それに、こちらで修正しますが、このエンドポイントではユーザ自身を削除できる訳で、それなのにエンドポイント名が `/profile` では、削除対象が合致しないという事態になっていましますね。
https://github.com/sls-training/2023-API-training-server/pull/1#discussion_r1186988938

tk0miya replied:
API を利用する側は、対象のアプリの考え方に習熟しているわけではないので、この API ドキュメントをみて「このアプリには `user` と `profile` という概念があるんだなあ」と解釈する可能性があります。

アプリケーションによってはユーザー (アカウント) とプロフィールが別だというものはいくつもあります。
たとえば Hulu や Netflix では、1つのアカウントに対して複数のプロフィールを紐付けることができます。

![](https://appllio.com/sites/default/files/styles/landscape_xl_1/public/2021/06/28/how-to-share-account-netflix-2021-06-001.jpg)

普段書くコードでも、同じデータのことをいくつもの呼び方をすると混乱が生じるように、API でもデータの呼び方は統一されていたほうがわかりやすいですし、誤解も減らせます。

この API を呼び出すクライアントアプリケーションは、この API ドキュメントを見て、API の呼び出し方を学ぶことになるので、自分の意図が伝えられるといいですね。
ですから、クライアントに対してこうしてほしい (こういう順序で呼び出してほしい、こういう状況ではこういうパラメータを渡してほしい) というのは明確に伝えるのが良いです。

a5ob7r replied:
> アプリケーションによってはユーザー (アカウント) とプロフィールが別だというものはいくつもあります。

ほほう。これは知らなかった。

a5ob7r replied:
安易に別名を使うと誤解を招きますね。でも、それはAPIに限らず何にでも言えることではありますね。気をつけるようにします。では、エンドポイント名は `/profile` から `/user` に修正します。

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+      summary: Upload a new file
+  /files/{fileId}:
+    get:
+      summary: Get a file info(file itself, name or description)
```
ID を指定しているので (ファイルを特定しているので) a じゃなくて the かなー

tk0miya commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+    parameters:
+      - name: fileId
+  /profile:
+    get:
+      summary: Get info of a logged-in user
```
ここって「任意の」ログインユーザーの情報を取る、ではなく「現在ログインしている」ユーザーの情報を取る、という意図ですよね、たぶん (サマリからは読み取れないので情報不足かも)
ということで、ここも a じゃなくて the かな。

tim-toga replied:
自分はあまり英語が得意では無いので、この辺の議論には積極的に参加できない。。
（可能なら日本語で書いてくれるとうれしいお気持ち）

tk0miya replied:
開発チーム内や、API の呼び出し元に対して意味や意図が正しく伝わるのが重要ですね。
OSS のように世界中の人に使ってもらう場合は英語で書いたほうが望ましいのですが、業務では海外とやりとりすることはレアケースなので日本語で書くほうが望ましいですね。海外の開発者とやり取りするとか、国際的なチームであると行った場合は別ですが、そういうケースは稀です。

他のところでコメントしましたが、API ドキュメントはクライアントアプリの開発者に対してこの API をどのように呼び出すのかを伝えるのが重要です。ですから

* 書いてある内容の正確性
* 情報に過不足がないこと
* (このアプリの知識が足りない人にも) API の呼び出し方や概念が伝わりやすいこと

が期待されます。

a5ob7r replied:
> ここって「任意の」ログインユーザーの情報を取る、ではなく「現在ログインしている」ユーザーの情報を取る、という意図ですよね、たぶん (サマリからは読み取れないので情報不足かも)
> ということで、ここも a じゃなくて the かな。

確かに。修正します。

a5ob7r replied:
> 自分はあまり英語が得意では無いので、この辺の議論には積極的に参加できない。。
> （可能なら日本語で書いてくれるとうれしいお気持ち）

英語で書いているのは、なんというか、習慣というか癖というか、ですね...。日本語化での記述は、どうしようかな...。

a5ob7r replied:
日本語化については別のプルリクで対応しようと考えています。

a5ob7r committed:
Remove some endpoints to send a form

The reason why is that WEB API users don't use forms to send requests
because APIs are opened and users know what parameters they need to send
to them. So providing APIs is instead of providing forms.

a5ob7r committed:
Add /signout instead of DELETE for /signin

This is more semantics.

tim-toga commented on openapi.yaml:
```diff
@@ -0,0 +1,43 @@
+  title: APIs to provide a simple online storage service
+  description: They are APIs to learn what APIs are and how to create them.
+  termsOfService: http://localhost:{PORT}/
+  version: 0.0.0
```
typo？version 0.0.0 になってる

a5ob7r replied:
まだ叩き台ですらない概要のみぐらいの意味合いで、わざと `0.0.0` にしてあります。違和感ありますかね。

tim-toga replied:
あ、であればOKです！

a5ob7r committed:
Fix a summary for DELETE for `/profile`

This API is just for a user deletion, not for logout.

a5ob7r committed:
Fix some definite articles in a document

a5ob7r committed:
Change `/profile` to `/user`

In general `profile` and `user` has different semantics on WEB APIs.

a5ob7r committed:
Move `/signup` endpoint into `/user`

No rational reason to separate `/user` into the two endpoints.

a5ob7r commented:
得られた承認が3つ、 [66ef160](https://github.com/sls-training/2023-API-training-server/pull/1/commits/66ef160ee6e2d7bff2e5d6b22d2348c6ed3f6d6c) での変更は開発者の裁量ということだったので、これでマージします。
